src/v/cloud_storage/manifest.cc

```diff
@@ -0,0 +1,442 @@
+    // for manifests, only 0x00000000, 0x10000000, ... 0xf0000000
+    // are used. This will allow us to quickly find all manifests
+    // that S3 bucket contains.
+    constexpr uint32_t bitmask = 0xF0000000;
+    auto path = ssx::sformat("{}_{}", ntp.path(), rev());
+    uint32_t hash = bitmask & xxhash_32(path.data(), path.size());
+    return remote_manifest_path(
+      fmt::format("{:08x}/meta/{}_{}/manifest.json", hash, ntp.path(), rev()));
+}
+
+remote_manifest_path manifest::get_manifest_path() const {
+    return generate_partition_manifest_path(_ntp, _rev);
+}
+
+remote_segment_path
+manifest::get_remote_segment_path(const segment_name& name) const {
+    auto path = ssx::sformat("{}_{}/{}", _ntp.path(), _rev(), name());
+    uint32_t hash = xxhash_32(path.data(), path.size());
+    return remote_segment_path(fmt::format("{:08x}/{}", hash, path));
+}
+
+const model::ntp& manifest::get_ntp() const { return _ntp; }
+
+const model::offset manifest::get_last_offset() const { return _last_offset; }
+
+model::revision_id manifest::get_revision_id() const { return _rev; }
+
+manifest::const_iterator manifest::begin() const { return _segments.begin(); }
+
+manifest::const_iterator manifest::end() const { return _segments.end(); }
+
+size_t manifest::size() const { return _segments.size(); }
+
+bool manifest::contains(const segment_name& obj) const {
+    return _segments.count(obj) != 0;
+}
+
+bool manifest::add(const segment_name& key, const segment_meta& meta) {
+    auto [it, ok] = _segments.insert(std::make_pair(key, meta));
+    _last_offset = std::max(meta.committed_offset, _last_offset);
+    return ok;
+}
+
+const manifest::segment_meta* manifest::get(const segment_name& key) const {
+    auto it = _segments.find(key);
+    if (it == _segments.end()) {
+        return nullptr;
+    }
+    return &it->second;
+}
+
+std::insert_iterator<manifest::segment_map> manifest::get_insert_iterator() {
+    return std::inserter(_segments, _segments.begin());
+}
+
+manifest manifest::difference(const manifest& remote_set) const {
+    if (_ntp != remote_set._ntp && _rev != remote_set._rev) {
+        throw std::logic_error(fmt_with_ctx(
+          fmt::format,
+          "{}-{} do not match {}-{}",
+          _ntp,
+          _rev,
+          remote_set._ntp,
+          remote_set._rev));
+    }
+    manifest result(_ntp, _rev);
+    std::set_difference(
+      begin(),
+      end(),
+      remote_set.begin(),
+      remote_set.end(),
+      result.get_insert_iterator());
+    return result;
+}
+
+ss::future<> manifest::update(ss::input_stream<char>&& is) {
+    using namespace rapidjson;
+    iobuf result;
+    auto os = make_iobuf_ref_output_stream(result);
+    co_await ss::copy(is, os);
+    iobuf_istreambuf ibuf(result);
+    std::istream stream(&ibuf);
+    Document m;
+    IStreamWrapper wrapper(stream);
+    m.ParseStream(wrapper);
+    update(m);
+    co_return;
+}
+
+void manifest::update(const rapidjson::Document& m) {
```

Comment:
if we expect errors, like parsing, should we encode it in `expect<error>`  with our boost-outcome wrapper? 

to me it read that we have exactly one version, and if this code cannot understand the version, i don't see how it can continue, unless there is some retry-loop above this.